### PR TITLE
Support referrerpolicy in amp-pixel.

### LINKED ISSUE #8372: Support referrer-policy for amp-pixel and amp-analytics
We should be able to support this for image requests, and possibly other transports.

Modern browsers support `referrerPolicy` directly, but we need to implement a fallback based on iframes with anonymous origins as a fallback. That means we can likely only support the value `none`.

### PR DESCRIPTION
**Pull request by lannka**
Closes #8372

Tested under latest Chrome, Safari, FF

### CODE REVIEW

File: builtins/amp-pixel.js [hunk 1]
```diff
@@ -69,9 +70,29 @@ export class AmpPixel extends BaseElement {
           .expandAsync(this.assertSource_(src))
           .then(src => {
             const image = new Image();
-            image.src = src;
-            dev().info(TAG, 'pixel triggered: ', src);
-            return image;
+            const referrerPolicy = this.element.getAttribute('referrerpolicy');
+            // if "referrerPolicy" is not supported, use iframe wrapper
```
Comment (cramforce) [inline builtins/amp-pixel.js]: Please factor into function.

File: builtins/amp-pixel.js [hunk 2]
```diff
@@ -69,9 +70,29 @@ export class AmpPixel extends BaseElement {
...
+            // to scrub the referrer.
+            if (image.referrerPolicy === undefined &&
+                referrerPolicy == 'no-referrer') {
```
Comment (cramforce) [inline builtins/amp-pixel.js]: This is the only value we support. That should be checked in the validator and be a user assert here. If the value is empty we can return early.

File: builtins/amp-pixel.js [hunk 3]
```diff
@@ -69,9 +70,29 @@ export class AmpPixel extends BaseElement {
...
+                  this.win.document, 'iframe', {
+                    src: `javascript: '<img src="${src}">'`,
+                  });
+              this.element.appendChild(iframe);
+              dev().info(TAG, 'pixel triggered via iframe: ', src);
+              return iframe;
+            } else {
+              if (referrerPolicy) {
+                image.referrerPolicy = referrerPolicy;
+                // referrerPolicy is respected only if the image is attached
```
Comment (cramforce) [inline builtins/amp-pixel.js]: Really? Is that a bug?

File: builtins/amp-pixel.md [hunk 4]
```diff
@@ -42,6 +42,11 @@ The `amp-pixel` component behaves like a simple tracking pixel `img`. It takes a
 ## Attributes
 
 **src** (required)
+The value is the URL of a remote endpoint.
+
+**referrerpolicy** (optional)
+Same as the `referrerpolicy` on `<img>`, see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy) for all valid values.
+For browsers don't support this attribute, `referrerpolicy=no-referrer` will trigger an iframe based solution to remove the `referrer` header in the HTTP request. Any other values will be a no-op.
```
Comment (cramforce) [inline builtins/amp-pixel.md]: As said above, we should only support the value we can support.

File: builtins/amp-pixel.js (hunk 1 above)
Comment (lannka) [inline builtins/amp-pixel.js]: done

File: builtins/amp-pixel.js (hunk 2 above)
Comment (lannka) [inline builtins/amp-pixel.js]: done

File: builtins/amp-pixel.js (hunk 3 above)
Comment (lannka) [inline builtins/amp-pixel.js]: behavior confirmed. 
but don't see it documented in the [spec](https://www.w3.org/TR/referrer-policy/#referrer-policy-delivery-referrer-attribute). could be a bug. not sure what to expect

File: builtins/amp-pixel.md (hunk 4 above)
Comment (lannka) [inline builtins/amp-pixel.md]: done

File: builtins/amp-pixel.js (hunk 3 above)
Comment (lannka) [inline builtins/amp-pixel.js]: take my word back. it was my mistake.

File: builtins/amp-pixel.js [hunk 5]
```diff
@@ -68,10 +69,14 @@ export class AmpPixel extends BaseElement {
       return urlReplacementsForDoc(this.element)
           .expandAsync(this.assertSource_(src))
           .then(src => {
-            const image = new Image();
-            image.src = src;
+            let pixel;
+            if (this.element.getAttribute('referrerpolicy') == 'no-referrer') {
+              pixel = createNoReferrerPixel(this.element, src);
+            } else {
+              pixel = createImagePixel(src);
```
Comment (cramforce) [inline builtins/amp-pixel.js]: Assert that this.element.getAttribute('referrerpolicy') is empty

File: builtins/amp-pixel.js [hunk 6]
```diff
@@ -68,10 +69,14 @@ export class AmpPixel extends BaseElement {
       return urlReplacementsForDoc(this.element)
           .expandAsync(this.assertSource_(src))
           .then(src => {
-            const image = new Image();
-            image.src = src;
+            let pixel;
+            if (this.element.getAttribute('referrerpolicy') == 'no-referrer') {
```
Comment (cramforce) [inline builtins/amp-pixel.js]: Comment why only this one.

File: test/functional/test-amp-pixel.js [hunk 7]
```diff
@@ -116,6 +116,21 @@ describes.realWin('amp-pixel', {amp: true}, env => {
           'https://pubads.g.doubleclick.net/activity;r=111');
     });
   });
+
+  it('should respect referrerpolicy=no-referrer', () => {
```
Comment (cramforce) [inline test/functional/test-amp-pixel.js]: Write integration test and verify no referrer on the server?

File: builtins/amp-pixel.js (hunk 6 above)
Comment (dvoytenko) [inline builtins/amp-pixel.js]: Why not also include origins and other values?

File: builtins/amp-pixel.js [hunk 8]
```diff
@@ -90,6 +95,46 @@ export class AmpPixel extends BaseElement {
...
+  }
+  image.src = src;
+  return image;
+}
+
+/**
+ * @returns {boolean}
+ */
+function isReferrerPolicySupported() {
+  return new Image().referrerPolicy !== undefined;
```
Comment (dvoytenko) [inline builtins/amp-pixel.js]: Can this instead be done via: `'referrerPolicy' in Image.prototype`?

File: builtins/amp-pixel.js [hunk 9]
```diff
@@ -90,6 +95,46 @@ export class AmpPixel extends BaseElement {
...
+    image.referrerPolicy = 'no-referrer';
+  }
+  image.src = src;
+  return image;
+}
+
+/**
+ * @returns {boolean}
+ */
+function isReferrerPolicySupported() {
```
Comment (dvoytenko) [inline builtins/amp-pixel.js]: Can you document where this is not the case right now?

File: test/functional/test-amp-pixel.js (hunk 7 above)
Comment (lannka) [inline test/functional/test-amp-pixel.js]: At this moment, I don't think we have such infrastructure ready to test if a request is received on local server. or maybe @rsimha-amp is working on something related?

File: builtins/amp-pixel.js (hunk 8 above)
Comment (lannka) [inline builtins/amp-pixel.js]: done

File: builtins/amp-pixel.js (hunk 9 above)
Comment (lannka) [inline builtins/amp-pixel.js]: done

File: builtins/amp-pixel.js (hunk 5 above)
Comment (lannka) [inline builtins/amp-pixel.js]: shall we leave the job to validator and fall back to normal pixel?

Comment (lannka) [inline builtins/amp-pixel.js]: thinking twice, assert here will prevent unnoticeable mistakes which could leak sensitive data.

File: test/functional/test-amp-pixel.js (hunk 7 above)
Comment (lannka) [inline test/functional/test-amp-pixel.js]: added a TODO for integration test.

**Comment by cramforce** [comment]
We do have testing infra for checking local HTTP requests. It is used by the `amp-form` tests. @mkhatib built it. CC @cvializ

**Comment by cramforce** [comment]
I'll insist on testing this. If our promise is "no referrer" we must actually test that.

**Comment by lannka** [comment]
@cramforce integration test added. PTAL.

File: builtins/amp-pixel.js [hunk 10]
```diff
@@ -95,6 +107,49 @@ export class AmpPixel extends BaseElement {
...
+ */
+function createNoReferrerPixel(parentElement, src) {
+  if (isReferrerPolicySupported()) {
+    return createImagePixel(src, true);
+  } else {
+    // if "referrerPolicy" is not supported, use iframe wrapper
+    // to scrub the referrer.
+    const iframe = createElementWithAttributes(
+        /** @type {!Document} */ (parentElement.ownerDocument), 'iframe', {
+          src: `javascript: '<img src="${src}">'`,
```
Comment (cramforce) [inline builtins/amp-pixel.js]: Question: Will this work in prod with our CSP preventing javascript URLs?

I don't think do. Can you create an `about:blank` frame instead and append the image pixel instead?

File: test/functional/test-amp-pixel.js (hunk 7 above)
Comment (cramforce) [inline test/functional/test-amp-pixel.js]: Remove TODO

Comment (lannka) [inline test/functional/test-amp-pixel.js]: already removed.

File: builtins/amp-pixel.js (hunk 10 above)
Comment (lannka) [inline builtins/amp-pixel.js]: Good catch. Fixed. Added a manual test.

I'm also considering changing the fixture of our integration test, so that CSP is always enforced. But I'd probably do in a separate PR.
